### DISCUSSION TITLE
How  does the repository context been added to the code completion generation process?

### DISCUSSION THREAD

**Discussion by ANYMS-A** [TabbyML/tabby#3183] [Q&A] [answered]
Hi, I am new to the TabbyML, after reading the doc, I got a question about the "repository" context about Tabby. In the document, it saids that :
"Tabby fetches the source code from the repository, parses it into an AST, and stores it in the index. During LLM inference, this context is utilized for code completion, as well as chat and search functionalities."

I assume this process is similar to the RAG and the code AST would be large and should take a lot of space of input tokens window of the LLM, then if I use a relatively small LLM with limited context window length (e.g. 1024 or 2048),  does Tabby only provide a partial part of repo context to the LLM during the inference?

**Comment by wsxiaoys** [comment, marked as answer]
> does Tabby only provide a partial part of repo context to the LLM during the inference?

Yes - only relevant context (defined as a combined score of BM25 and semantic relevant) will be inserted into code completion context. See https://github.com/TabbyML/tabby/issues/2674 for a more complete view

**Reply by ANYMS-A** [reply]
Thanks ! That's really helpful
